src/redis/cache_impl.go

```diff
@@ -99,20 +117,47 @@ func (this *rateLimitCacheImpl) DoLimit(
 		current := uint32(conn.PipeResponse().Int())
 		conn.PipeResponse() // Pop off EXPIRE response and check for error.
 
+		limit := limits[i].Limit.RequestsPerUnit
+
+		// previous is the the value of the limit before adding addNHits.
+		previous := current - addNHits
+
+		// The nearLimitValue is the number of requests that can be made before hitting the NearLimitRatio.
+		// We need to know it in both the OK and OVER_LIMIT scenarios.
+		nearLimitValue := uint32(math.Floor(float64(float32(limit) * config.NearLimitRatio)))
+
 		logger.Debugf("cache key: %s current: %d", cacheKey, current)
-		if current > limits[i].Limit.RequestsPerUnit {
+		if current > limit {
 			responseDescriptorStatuses[i] =
 				&pb.RateLimitResponse_DescriptorStatus{pb.RateLimitResponse_OVER_LIMIT,
 					limits[i].Limit, 0}
+
+			// Increase over limit statistics. Because we support += behavior for increasing the limit, we need to
+			// asses if the entire addNHits were over the limit. That is if the limit's value before adding the
+			// N hits was over the limit, then all the N hits were over limit.
+			// Otherwise, only the difference between the current value and the limit were over limit hits.
+			if previous >= limit {
```

Comment:
I'm sorry but my brain hurts trying to make sense of the logic/math here. Can you try renaming variables to be more clear per some other comments and then I can look again.